internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -423,6 +427,21 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		if rdy {
 			ready++
 		}
+
+		if err := r.composite.Render(ctx, cr, cd, comp.Spec.Resources[i]); err != nil {
+			log.Debug(errRenderCR, "error", err)
+			r.record.Event(cr, event.Warning(reasonCompose, err))
+			return reconcile.Result{RequeueAfter: shortWait}, nil
+		}
+	}
+
+	// Update composite with patches from composed resources
+	// DeepCopy composite to preserve rendered status
+	crSpecObject := cr.DeepCopyObject().(client.Object)
+	if err := r.client.Update(ctx, crSpecObject); err != nil {
+		log.Debug(errUpdate, "error", err)
+		r.record.Event(cr, event.Warning(reasonCompose, err))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
 	}
```

Comment:
This seems reasonable, after more thought I'm curious what your opinions are on overwriting the `resourceVersion` here is. The spec is already updated here, so we are not forcing changes on legitimate changes, the `resourceVersion` overwrite would just be against any other status changes, which the controller should already own.

Other than overwriting, the early exit makes sense as another option to remove unneeded logging/events.

Reply:
I (very) slightly prefer the early exit because it eliminates the small chance we'd accidentally reset any status changes that had been persisted to the API server (by another party) if we overwrite the resource version.

For example:

1. We make a deepcopy.
2. Some other entity updates the XR's status.
3. We update the spec and metadata using our deepcopy.
4. We copy the deepcopy's resource version to our original object
5. We update the status using our (now stale) original object, thus overwriting the status changes.

That all said, I hadn't considered until you mentioned it that only status would be affected and in practice we're the only controller that updates status - so this is much less of a big deal than resetting a spec update for example. We do have some instances of multiple controllers updating status (e.g. two different controllers update XRD status) but it is admittedly quite an edge case.

I'd typically optimise for simplicity and readability over correctness in cases like this in which the correctness issue is so trivial, but to my mind both options are of similar complexity/readability.

Reply:
Makes sense, early exit is what I've gone with here after your suggestion 👍